tests/test_coordinate_source.py

```diff
@@ -0,0 +1,328 @@
+import os
+import pickle
+import random
+import shutil
+import tempfile
+from collections import OrderedDict
+from glob import glob
+from unittest import TestCase
+
+import importlib_resources
+import mrcfile
+import numpy as np
+from click.testing import CliRunner
+from pandas import DataFrame
+
+import tests.saved_test_data
+from aspire.commands.extract_particles import extract_particles
+from aspire.noise import WhiteNoiseEstimator
+from aspire.source.coordinates import (
+    CentersCoordinateSource,
+    EmanCoordinateSource,
+    RelionCoordinateSource,
+)
+from aspire.storage import StarFile
+
+
+class CoordinateSourceTestCase(TestCase):
+    def setUp(self):
+        # temporary directory to set up a toy dataset
+        self.tmpdir = tempfile.TemporaryDirectory()
+        self.data_folder = self.tmpdir.name
+        # load pickled list of picked particle centers
+        with importlib_resources.path(
+            tests.saved_test_data, "apple_centers.p"
+        ) as centers_path:
+            # apple_centers.p was pickled with protocol 4
+            centers = pickle.load(
+                open(str(centers_path), "rb"),
+            )
+        # get path to test .mrc file
+        with importlib_resources.path(tests.saved_test_data, "sample.mrc") as test_path:
+            self.original_mrc_path = str(test_path)
+        # save test data root dir
+        self.test_dir_root = os.path.dirname(self.original_mrc_path)
+
+        # First set up tests for EmanCoordinateSource and CentersCoordinateSource
+        # We will construct a source with two micrographs and two coordinate
+        # files by using the same micrograph, but dividing the coordinates
+        # among two files (this simulates a dataset with multiple micrographs)
+        for i in range(2):
```

Comment:
Lot of things happening here. Maybe split some parts off to functions? (Like writing the box file, which we use twice.)

Reply:
Added separate functions to construct each filetype